### ISSUE TITLE
Conbee not detected when another USB device is connected

### ISSUE THREAD

**Issue by geelenbert** [marthoc/hassio-addons#2]
I’m trying to get the HASS io plugin working. How do i specify which USB device to use?

I have a conbee connected to hass. It shows up in the hardware listing when i SSH into HASS and give the command “hassio host hardware” .

I also have other Serial usb devices connected to my hass io (smart meter reading).

If i go to the webpage, i only see a conbee gateway with the ip address hass IO, no conbee device as i previous had on an other system running deconz.

After some experimenting i found the source.
It works only if the deconz usb device is the only usb serial device connected to hass io.
Any idea how to do a work around ?

My suggestion would be to use a conif paramter with the usb name of the conbee gateway. I would use the /dev/by-id/ device path to always point to the conbee gateway.

Device name should not be fixed to dev/TTYUSB0 , as the number of the USB number can change due to the number of devices plugged in.

**Comment by marthoc** [comment]
Currently I’m using the `auto_uart` feature of the Hass.io addon config, which is supposed to pass through all devices to the addon container. Clearly it’s not allowing deCONZ to detect the Conbee properly in this case.

I suppose I will have to look into converting this into a config parameter for the serial device and abandoning the auto_uart approach.

**Comment by ggravlingen** [comment]
I had the same problem when enabling Bluetooth of my Raspberry pi 3 (and solved it by disabling BT again.) I havent tried it myself, but some post suggested to set `enable_uart=1` in the `/boot/config.txt` file.

**Comment by marthoc** [comment]
Yes that line gets added when you follow the steps in the README regarding disabling the login shell over serial and enabling the serial port (UART=serial port). I happen to know that OP is not on a RPi however so I think the problem remains. I’m wondering if it’s related to this deCONZ issue: https://github.com/dresden-elektronik/deconz-rest-plugin/issues/511.

**Comment by geelenbert** [comment]
Indeed i'm running HASS inside a VM, not on an Raspberry. 

I updated the firmware and deconz to the latest version, but does not help. 

Here maybee some detailed info:
- I boot with Conbee and other USB device connected, Conbee gets /dev/ttyUSB1 assigned. Deconz can't find the USB stick.
- I disconnect the other USB device from the VM (USB0), reboot Deconz, still no connection. The Conbee is still USB1 at this point.
- I disconnect the other USB device from the VM, reboot HASS, Deconz connects to the Conbee. Conbee us USB0

Hope this helps narrowing it down.
let me know how i can help.

**Comment by marthoc** [comment]
Out of interest, what if Conbee is USB0 and you plug in the other stick then start deconz? Does it work or not?

**Comment by geelenbert** [comment]
@marthoc yes it works. After restarting the container Deconz still works. As soon as Conbee is nog USB0, it looks like it goes wrong....

**Comment by marthoc** [comment]
@geelenbert So I think this is a current known problem of Hass.io where it can't guarantee consistent device names. It does seem to be possible to set up a udev rule on the underlying os (Resinos) - can you try to do this and set it so that Conbee is always assigned /dev/ttyUSB0? I think this may fix the problem in the short term until HassioOS is released (not sure when that will be though).

**Comment by geelenbert** [comment]
@marthoc Not sure if that is possible. I know you can create symlinks using UDEV rules to point to dev/ttyUSB0 ( fox example, i create a rule that creates a link from /dev/ttyCONBEEUSB -> dev/ttyUSB0)

/dev/ttyUSB0 is created by the OS, so i'd rather not mess with that. 

But i don't really see where the issue is here. 
if i look at your deconz docker properties, there is a property you use:
https://github.com/marthoc/docker-deconz
```
docker run -d \
    --name=deconz \
    --net=host \
    --restart=always \
    -v /opt/deconz:/root/.local/share/dresden-elektronik/deCONZ \
    --device=/dev/ttyUSB0 \
    marthoc/deconz
```

Looks like just this parameter needs to be configurable

**Comment by geelenbert** [comment]
@marthoc Can you point me to the earlier discussion on the Deconz app where there was talk about the command line paramters ? I will start there.

**Comment by marthoc** [comment]
@geelenbert - you can see some discussion about the command line parameter in this Issue: https://github.com/dresden-elektronik/deconz-rest-plugin/issues/236. @manup (the main deCONZ developer) says that a command-line switch to specify the device is "on the list" but maybe we can get an update from him on that feature. 😄

**Comment by kronnberg** [comment]
@marthoc I'm having a similar problem with my conbee. It seems that it's not detected despite being the only usb device connected. hassio host hardware shows that its connected to /dev/ttyUSB0. Looking at the log it seems like it tries to connect to /dev/ttyAMA0.

17:56:58:522 void zmMaster::handleStateIdle(zmMaster::MasterEvent) not connected goto OFF state
17:56:58:522 device state timeout ignored in state 1
17:56:59:439 dev /dev/ttyAMA0
17:56:59:983 device state timeout ignored in state 2
17:57:01:022 device state timeout ignored in state 2
17:57:02:022 device state timeout ignored in state 2
17:57:03:023 device state timeout ignored in state 2
17:57:04:023 device state timeout ignored in state 2
17:57:04:523 Daylight now: solarNoon, status: 170
17:57:05:022 device state timeout (handled)

What can I do/try? my linux/docker knowledge is very limited.

**Comment by ggravlingen** [comment]
@kronnberg try disabling Bluetooth and see if it works.

**Comment by kronnberg** [comment]
@ggravlingen Ok
Is it enough to set 
`dtoverlay=pi3-disable-bt` in /boot/config.txt ?

**Comment by kronnberg** [comment]
@ggravlingen That did the trick! It works now. Thank you!

**Comment by Erlandgs** [comment]
@marthoc I also have problems getting the addon to find my conbee stick, in spite of it being the only device connected. I have tried disabling bluetooth as described by @kronnberg but it did not help. I'm new to this so I also don't really know where to find the logs that @kronnberg described to actually see if I have the same problem. But the following serial devices are detected (after disabling bluetooth):
/dev/ttyAMA0
/dev/ttyUSB0
/dev/ttyS0
I'm using a raspberry pi zero, not sure if this changes anything? 
The conbee stick works fine on my rpi3 running full home assistant, and I have updated its fw to latest version.

**Comment by ggravlingen** [comment]
I was never able to get the conbee to work on a pi zero :( Is the processing capacity enough on the device.

**Comment by marthoc** [comment]
@Erlandgs - I have no experience with the Pi Zero unfortunately but the fact that /dev/ttyAMA0 is still present in the system means that some serial device is still active - are you using Raspbian or Hass.io? If Raspbian, have you followed the instructions in the README about disabling the serial port login, etc?

Unfortunately I think the deCONZ program on arm looks for /dev/ttyAMA0 first, and if it finds one it seems to hang and timeout if it’s not a RaspBee.

**Comment by Erlandgs** [comment]
Thanks for fast answers!

I'm using hass.io. I tried disabling /dev/ttyAMA0 by adding the following to the config.txt:
`enable_uart=0
dtoverlay=pi3-miniuart-bt`

This resulted in only the following serial hw being detected:
/dev/ttyUSB0
/dev/ttyS0

But it still does not work. 
The hass.io system logs (from the web UI: Hass.io -> system) are not so helpful. 
`18-04-29 13:41:36 INFO (SyncWorker_0) [hassio.docker.addon] Start docker addon marthoc/hassio-addon-deconz-armhf with version 2.05.23`
Are there more helpful logs? I have added the debug logging to the configuration.yaml according to [https://www.home-assistant.io/components/deconz/](https://www.home-assistant.io/components/deconz/). But I'm guessing this is for logging once the add-on is working?

**Comment by geelenbert** [comment]
> 
> 
> @geelenbert - you can see some discussion about the command line parameter in this Issue: [dresden-elektronik/deconz-rest-plugin#236](https://github.com/dresden-elektronik/deconz-rest-plugin/issues/236). @manup (the main deCONZ developer) says that a command-line switch to specify the device is "on the list" but maybe we can get an update from him on that feature. smile

It looks like the desired function is added in Deconz:
https://github.com/dresden-elektronik/deconz-rest-plugin/releases/tag/V2_05_31

- New deCONZ command line parameter --dev=<com device> to connect to a specific device like /dev/ttyUSB0

**Comment by marthoc** [comment]
I've pushed a new version of the add-on that now requires you to specify the deCONZ device name in the config. This leverages the --dev option that was added to deCONZ and should solve this problem. Please test and report back if working or not.
